### PR TITLE
[Merged by Bors] - feat: port CategoryTheory.Shift.Basic

### PR DESCRIPTION
**Pull request by joelriou**
---
- [x] depends on: #2466
- [x] depends on: #3038
- [x] depends on: leanprover-community/mathlib#18670

While doing the port, in order to un-obfuscate some proofs involving shifts, I have tried to try avoid `MonoidalCategory` API leakage when using `HasShift`. I feel that the abbreviations which were part of mathlib3 `Shift` API should be defs, and that all the relevant lemmas should be stated using these definitions instead. Then, for example, when we deal with shifts in triangulated categories and we have to show a compatibility, we should not always be reminded that shifts were implemented using monoidal categories. Then, when we need a new compatibility which requires unfolding the definitions, we may add it to `Shift.lean`. Indeed, I have added suitable lemmas to `Shift.lean` so that almost all obfuscated proofs in `Triangulated.Rotate` are just `by aesop_cat`, see #3047

This is backported in https://github.com/leanprover-community/mathlib/pull/18670 which is now merged.

I shall wait for `mathport` output in order to update sha and #align statements. Done.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2466
* leanprover-community/mathlib4#3038
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2466~~
* ~~leanprover-community/mathlib4#3038~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by ChrisHughes24** [comment]
@joelriou Is this awaiting-review?

**Comment by Parcly-Taxel** [comment]
> @ChrisHughes24 I am not sure in which order this should be reviewed: first the mathlib refactor PR [#18670](https://github.com/leanprover-community/mathlib/pull/18670), or this one?

I'd go with #18670 first; the mathlib3 PR eases the whole chain of mathlib4 PRs you have.

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2466~~
* ~~leanprover-community/mathlib4#3038~~
* ~~leanprover-community/mathlib#18670~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/CategoryTheory/Shift/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,760 @@
...
+* `HasShift`: A typeclass asserting the existence of a shift functor.
+* `shiftEquiv`: When the indexing monoid is a group, then the functor indexed by `n` and `-n` forms
+  an self-equivalence of `C`.
+* `shiftComm`: When the indexing monoid is commutative, then shifts commute as well.
+
+## Implementation Notes
+
+`[HasShift C A]` is implemented using `MonoidalFunctor (Discrete A) (C ⥤ C)`.
+However, the API of monodial functors is used only internally: one should use the API of
+shifts functors which includes `shiftFunctor C a : C ⥤ C` for `a : A`,
```
Comment (semorrison) [inline Mathlib/CategoryTheory/Shift/Basic.lean:34]:
```suggestion
shift functors which includes `shiftFunctor C a : C ⥤ C` for `a : A`,
```

**Comment by semorrison** [comment]
Thank you so much, @joelriou, for your attention to shift functions here and in mathlib3. I had been very sad about the state we'd left them in, and I'm hopeful that you've resolved our problems!

bors merge

**Pull request closed without merging**